pkg/pprof/cell.go

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package pprof
+
+import (
+	"errors"
+	"net"
+	"net/http"
+	_ "net/http/pprof"
+	"strconv"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/pflag"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/hive/cell"
+)
+
+const (
+	// Pprof is the flag to enable the registration of pprof HTTP handlers
+	Pprof = "pprof"
+
+	// PprofAddress is the flag to set the address that pprof listens on
+	PprofAddress = "pprof-address"
+
+	// PprofPort is the flag to set the port that pprof listens on
+	PprofPort = "pprof-port"
+)
+
+// Cell creates the cell for pprof, that registers its HTTP handlers to serve
+// profiling data in the format expected by the pprof visualization tool.
+var Cell = cell.Module(
+	"pprof",
+	"pprof HTTP handlers",
+
+	// We don't call cell.Config directly here, because the operator
+	// uses different flags names for the same pprof config flags.
+	// Therefore, to register each flag with the correct name, we leave
+	// the call to cell.Config to the user of the cell.
+
+	// Invoke forces the registration of the pprof hooks
+	cell.Invoke(registerPprofHooks),
+)
+
+// Config contains the configuration for the pprof cell.
+type Config struct {
+	Pprof        bool
+	PprofAddress string
+	PprofPort    int
+}
+
+func (def Config) Flags(flags *pflag.FlagSet) {
+	flags.Bool(Pprof, def.Pprof, "Enable serving pprof debugging API")
+	flags.String(PprofAddress, def.PprofAddress, "Address that pprof listens on")
+	flags.Int(PprofPort, def.PprofPort, "Port that pprof listens on")
+}
+
+func registerPprofHooks(lc hive.Lifecycle, log logrus.FieldLogger, cfg Config) {
+	if !cfg.Pprof {
+		return
+	}
+
+	addr := net.JoinHostPort(cfg.PprofAddress, strconv.Itoa(cfg.PprofPort))
+	log = log.WithFields(logrus.Fields{"address": addr})
+	srv := &http.Server{Addr: addr}
+	lc.Append(hive.Hook{
+		OnStart: func(ctx hive.HookContext) error {
+			log.Info("Started pprof server")
+			go func() {
+				if err := srv.ListenAndServe(); !errors.Is(err, http.ErrServerClosed) {
+					log.WithError(err).Error("server stopped unexpectedly")
```

Comment:
Exactly, that was my thought.
I see that for [metrics server](https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_main.go#L1868) we call `Shutdowner.Shutdown`, but I think that for pprof it is reasonable to leave out the error, apart from logging it.

Reply:
agreed, since these are host-network pods, the chance of a port collision is non-zero. And it really sucks to lose networking because of that. Better to log and continue.